hpy/debug/src/_debugmod.c

```diff
@@ -0,0 +1,212 @@
+*/
+
+typedef struct {
+    HPyObject_HEAD
+    DebugHandle *handle;
+} DebugHandleObject;
+
+HPyDef_GET(DebugHandle_obj, "obj", DebugHandle_obj_get,
+           .doc="The object which the handle points to")
+static UHPy DebugHandle_obj_get(HPyContext uctx, UHPy self, void *closure)
+{
+    DebugHandleObject *dh = HPy_CAST(uctx, DebugHandleObject, self);
+    return HPy_Dup(uctx, dh->handle->uh);
+}
+
+HPyDef_GET(DebugHandle_id, "id", DebugHandle_id_get,
+           .doc="A numeric identifier representing the underlying universal handle")
+static UHPy DebugHandle_id_get(HPyContext uctx, UHPy self, void *closure)
+{
+    DebugHandleObject *dh = HPy_CAST(uctx, DebugHandleObject, self);
+    return HPyLong_FromSsize_t(uctx, (HPy_ssize_t)dh->handle);
+}
+
+HPyDef_SLOT(DebugHandle_cmp, DebugHandle_cmp_impl, HPy_tp_richcompare)
+static UHPy DebugHandle_cmp_impl(HPyContext uctx, UHPy self, UHPy o, HPy_RichCmpOp op)
+{
+    UHPy T = HPy_Type(uctx, self);
+    if (!HPy_TypeCheck(uctx, o, T))
+        return HPy_Dup(uctx, uctx->h_NotImplemented);
+    DebugHandleObject *dh_self = HPy_CAST(uctx, DebugHandleObject, self);
+    DebugHandleObject *dh_o = HPy_CAST(uctx, DebugHandleObject, o);
+
+    switch(op) {
+    case HPy_EQ:
+        return HPyBool_FromLong(uctx, dh_self->handle == dh_o->handle);
+    case HPy_NE:
+        return HPyBool_FromLong(uctx, dh_self->handle != dh_o->handle);
+    default:
+        return HPy_Dup(uctx, uctx->h_NotImplemented);
+    }
+}
+
+HPyDef_SLOT(DebugHandle_repr, DebugHandle_repr_impl, HPy_tp_repr)
+static UHPy DebugHandle_repr_impl(HPyContext uctx, UHPy self)
+{
+    DebugHandleObject *dh = HPy_CAST(uctx, DebugHandleObject, self);
+    UHPy uh_fmt = HPy_NULL;
+    UHPy uh_id = HPy_NULL;
+    UHPy uh_args = HPy_NULL;
+    UHPy uh_result = HPy_NULL;
+
+    uh_fmt = HPyUnicode_FromString(uctx, "<DebugHandle 0x%x for %r>");
+    if (HPy_IsNull(uh_fmt))
+        goto exit;
+
+    uh_id = HPyLong_FromSsize_t(uctx, (HPy_ssize_t)dh->handle);
+    if (HPy_IsNull(uh_id))
+        goto exit;
+
+    uh_args = HPyTuple_FromArray(uctx, (UHPy[]){uh_id, dh->handle->uh}, 2);
+    if (HPy_IsNull(uh_args))
+        goto exit;
+
+    uh_result = HPy_Remainder(uctx, uh_fmt, uh_args);
+
+ exit:
+    HPy_Close(uctx, uh_fmt);
+    HPy_Close(uctx, uh_id);
+    HPy_Close(uctx, uh_args);
+    return uh_result;
+}
+
+
+
+static HPyDef *DebugHandleType_defs[] = {
+    &DebugHandle_obj,
+    &DebugHandle_id,
+    &DebugHandle_cmp,
+    &DebugHandle_repr,
+    NULL
+};
+
+static HPyType_Spec DebugHandleType_spec = {
+    .name = "hpy.debug._debug.DebugHandle",
+    .basicsize = sizeof(DebugHandleObject),
+    .flags = HPy_TPFLAGS_DEFAULT,
+    .defines = DebugHandleType_defs,
+};
+
+
+static UHPy new_DebugHandleObj(HPyContext uctx, UHPy u_DebugHandleType,
+                               DebugHandle *handle)
+{
+    DebugHandleObject *dhobj;
+    UHPy u_result = HPy_New(uctx, u_DebugHandleType, &dhobj);
```

Comment:
Would it make sense to put the DebugHandle type on the context like we do for other build-in types like List?

Reply:
I guess this is also a good illustration of the slight hopping one has to do at the moment since global type objects are gone -- and the hopping is actually not too bad in simple cases.

Reply:
> Would it make sense to put the DebugHandle type on the context like we do for other build-in types like List?

no, because it's not a builtin. Why should it be different than, say, `ndarray`?

> I guess this is also a good illustration of the slight hopping one has to do at the moment since global type objects are gone -- and the hopping is actually not too bad in simple cases.

exactly, that's the point. We need a way to have per-module "globals" at the C level, which we still don't have :(